File: sdk/python/src/graft/modules/oled.py
"""GRAFT OLED SSD1306 Module"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from graft.device import GraftDevice

from graft.core.constants import CMD


class GraftOLED:
    def __init__(self, device: GraftDevice, module_id: int = 0) -> None:
        self._device = device
        self._mod_id = module_id

    async def begin(self, width: int = 128, height: int = 64, i2c_addr: int = 0x3C) -> None:
        payload = bytes([self._mod_id, 0x01, width, height, i2c_addr])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def clear(self) -> None:
        payload = bytes([self._mod_id, 0x02])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def text(self, x: int, y: int, size: int, text: str) -> None:
        text_bytes = text.encode("utf-8")
        payload = bytes([self._mod_id, 0x03, x, y, size]) + text_bytes
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def draw_rect(self, x: int, y: int, w: int, h: int, fill: bool = False) -> None:
        payload = bytes([self._mod_id, 0x04, x, y, w, h, 1 if fill else 0])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def draw_bitmap(self, x: int, y: int, w: int, h: int, data: bytes) -> None:
        payload = bytes([self._mod_id, 0x05, x, y, w, h]) + data
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def show(self) -> None:
        payload = bytes([self._mod_id, 0x06])
        await self._device._send_command(CMD.MOD_CMD, payload)

File: sdk/js/src/modules/oled.ts
/**
 * GRAFT OLED SSD1306 Module — Typed JS wrapper
 */

import type { GraftDevice } from '../device.js'

export class GraftOLED {
  private _device: ReturnType<GraftDevice['module']>

  constructor(device: GraftDevice) {
    this._device = device.module('oled1306')
  }

  /** Initialize the OLED display. */
  async begin(width = 128, height = 64, i2cAddr = 0x3C): Promise<void> {
    await this._device.cmd(0x01, new Uint8Array([width, height, i2cAddr]))
  }

  /** Clear the display buffer. Call show() to apply. */
  async clear(): Promise<void> {
    await this._device.cmd(0x02)
  }

  /** Draw text at position with size. */
  async text(x: number, y: number, size: number, str: string): Promise<void> {
    const strBytes = new TextEncoder().encode(str)
    const buf = new Uint8Array(3 + strBytes.length)
    buf[0] = x; buf[1] = y; buf[2] = size
    buf.set(strBytes, 3)
    await this._device.cmd(0x03, buf)
  }

  /** Draw a rectangle. fill=true for filled. */
  async drawRect(x: number, y: number, w: number, h: number, fill = false): Promise<void> {
    await this._device.cmd(0x04, new Uint8Array([x, y, w, h, fill ? 1 : 0]))
  }

  /** Draw a 1-bit bitmap. */
  async drawBitmap(x: number, y: number, w: number, h: number, data: Uint8Array): Promise<void> {
    const buf = new Uint8Array(4 + data.length)
    buf[0] = x; buf[1] = y; buf[2] = w; buf[3] = h
    buf.set(data, 4)
    await this._device.cmd(0x05, buf)
  }

  /** Flush the display buffer to the screen. */
  async show(): Promise<void> {
    await this._device.cmd(0x06)
  }
}
